### ISSUE TITLE
Support for S3 events?

### ISSUE THREAD

**Issue by moravcik** [ar90n/serverless-s3-local#13]
I haven't tried yet, but does this support also S3 events for lambdas?

**Comment by ar90n** [comment]
Hi @moravcik 
serverless-s3-local doesn't support S# events for lambdas.

**Comment by n1ru4l** [comment]
Hey there I am one of the new maintainers of s3rver. We got an ongoing PR for that @moravcik. Would love to integrate this somehow in this library once it is merged 😊

**Comment by ar90n** [comment]
Hi @n1ru4l 
Thanks for your good news !
@moravcik 's PR sounds good for this project.
I'm going to research how to integrate with it.
I can't wait for the PR will be merged.

**Comment by n1ru4l** [comment]
Hey there! Just wanted to let you know that the feature is now in the master branch and will be released soon ☺️

**Comment by n1ru4l** [comment]
@moravcik FYI I am talking about s3rver not this library, we will have to figure out on how to glue it together

**Comment by ar90n** [comment]
@moravcik @n1ru4l Thank for good news! I just started to develop.

**Comment by n1ru4l** [comment]
Some ideas:

Detect if `serverless-offline-sns` is installed and if so propagate events emitted by `s3rver.client.subscribe`.

https://github.com/jamhall/s3rver#subscribing-to-s3-event

**Comment by ar90n** [comment]
@n1ru4l Thanks for your suggestion. serverless-offline-sns seems great example for me. But it seems only suitable for SNS event. So I decided to implement the function to call lambda by referring it.

**Comment by n1ru4l** [comment]
Oh I did not know you can use S3 directly as an Event Source. Makes  sense to do it like you did. Looking forward to use it once it is shipped 😊!

**Comment by ar90n** [comment]
@moravcik @n1ru4l I have just implemented s3 event support at support-s3-event branch.
If you don't mind, please see the example/simple_event.
It is simple demo for this function.
It runs the following.

Server terminal
```
$ cd example/simple_event
$ sls offline
Serverless: Found S3 event listener for local-bucket
S3 local started ( port:8000 )
Serverless: creating bucket: local-bucket
info: [S3rver] Created new bucket "local-bucket" successfully
info: [S3rver] PUT /local-bucket 200 - - 10.048 ms
Serverless: Starting Offline: dev/us-east-1.

Serverless: Routes for webhook:
Serverless: GET /

Serverless: Routes for s3hook:
Serverless: (none)

Serverless: Offline listening on http://localhost:3000

Serverless: GET / (λ: webhook)
Serverless: The first request might take a few extra seconds
Serverless: [200] "ok"
info: [S3rver] Stored object "1234" in bucket "local-bucket" successfully
{"Records":[{"eventVersion":"2.0","eventSource":"aws:s3","awsRegion":"us-east-1","eventTime":"2018-02-24T18:02:22.077Z","eventName":"ObjectCreated:Put","userIdentity":{"principalId":"AWS:F5910C505B67C4D363CB1"},"requestParameters":{"sourceIPAddress":"127.0.0.1"},"responseElements":{"x-amz-request-id":"9064BDBFFC3367C1","x-amz-id-2":"WviFkakBm0CFCVbyeIKQwv5OOD5sBWGTTzv9gA+b3S8="},"s3":{"s3SchemaVersion":"1.0","configurationId":"testConfigId","bucket":{"name":"local-bucket","ownerIdentity":{"principalId":"039920F2F15A5B"},"arn":"arn:aws:s3: : :local-bucket"},"object":{"key":"1234","sequencer":"161C8FAE805","size":4,"eTag":"e2fc714c4727ee9395f324cd2e7f331f"}}}]}
{"functionName":"serverless-s3-local-example-dev-s3hook","functionVersion":"offline_functionVersion_for_serverless-s3-local-example-dev-s3hook","invokedFunctionArn":"offline_invokedFunctionArn_for_serverless-s3-local-example-dev-s3hook","awsRequestId":"offline_awsRequestId_3100985216280683","logGroupName":"offline_logGroupName_for_serverless-s3-local-example-dev-s3hook","logStreamName":"offline_logStreamName_for_serverless-s3-local-example-dev-s3hook","identity":{},"clientContext":{}}
info: [S3rver] PUT /local-bucket/1234 200 - - 11.471 ms
info: [S3rver] Stored object "README.md" in bucket "local-bucket" successfully
{"Records":[{"eventVersion":"2.0","eventSource":"aws:s3","awsRegion":"us-east-1","eventTime":"2018-02-24T18:02:34.449Z","eventName":"ObjectCreated:Put","userIdentity":{"principalId":"AWS:309681C52386ACA930562"},"requestParameters":{"sourceIPAddress":"127.0.0.1"},"responseElements":{"x-amz-request-id":"69EADA24D7FD3610","x-amz-id-2":"WviFkakBm0CFCVbyeIKQwv5OOD5sBWGTTzv9gA+b3S8="},"s3":{"s3SchemaVersion":"1.0","configurationId":"testConfigId","bucket":{"name":"local-bucket","ownerIdentity":{"principalId":"D6A3928B3B6AAB"},"arn":"arn:aws:s3: : :local-bucket"},"object":{"key":"README.md","sequencer":"161C8FB1859","size":49,"eTag":"3025a9a94ac1c7ec794fc1a10bec2705"}}}]}
{"functionName":"serverless-s3-local-example-dev-s3hook","functionVersion":"offline_functionVersion_for_serverless-s3-local-example-dev-s3hook","invokedFunctionArn":"offline_invokedFunctionArn_for_serverless-s3-local-example-dev-s3hook","awsRequestId":"offline_awsRequestId_3100985216280683","logGroupName":"offline_logGroupName_for_serverless-s3-local-example-dev-s3hook","logStreamName":"offline_logStreamName_for_serverless-s3-local-example-dev-s3hook","identity":{},"clientContext":{}}
info: [S3rver] PUT /local-bucket/README.md 200 - - 12.152 ms
```

Client terminal
```
$ curl http://localhost:3000/
$ aws --endpoint http://localhost:8000 s3api put-object --bucket local-bucket --key README.md --body ./README.md
{
    "ETag": "\"3025a9a94ac1c7ec794fc1a10bec2705\""
}
```

Is it correct for your request ?

**Comment by floo51** [comment]
Hello there !
Much interested by this feature, reviewed the code and examples on the `support-s3-event` branch and it seems to do the job well.
Any help needed with testing or documenting ?

**Comment by ar90n** [comment]
Hi @floo51 
Thanks for your comment.
I'm happy because you have been interested in this project.
I'm waiting for your contribution in every way.
For testing,  please use this function in your weekend hobby project and report issues and requests.
For documenting, I want every helps because I have only poor english. If your don't mind, please modify our README.md and comments.

**Comment by floo51** [comment]
For what I tested, there's an issue when using `serverless-webpack` plugin.
We try to declare handlers from routes before the webpack output is built, makes it crash at startup.

I will try to see how they do it in `serverless-offline` and submit a pull request on your `support-s3-event` branch.

Once it's done, I will check documentation as well !

**Comment by ar90n** [comment]
@floo51 
Thanks for your information.
I'm looking forward to your PR !!

**Comment by jml6m** [comment]
Would love to have this functionality - any updates on when it will be released?

**Comment by ar90n** [comment]
I decided to merge `support-s3-event` branch to master. 
Because this function may make someone benefits. 

@jml6m 
Thanks for your comment.
I merged and published this function as version 0.3.0.
I wish this project will help you.

@floo51 
For now, I'm going to close this thread.
I'm waiting for your contribution with pleasure.
In that case, please create new issue or PR  for your contribution.

**Comment by anantab** [comment]
Thanks @ar90n for adding the feature in the library. I needed the feature in this library, so I added s3Event feature  in S3rver library first. I was thinking of doing PR of this feature but you already have done it. Thanks.

**Comment by ar90n** [comment]
Hi @anantab 
Thanks for your comment.
Your great work in s3rver made me motivated.
If this function help you, I will be happy.
